RecoLocalTracker/SiPhase2VectorHitBuilder/src/VectorHitBuilderAlgorithm.cc

```diff
@@ -0,0 +1,574 @@
+  lowerClusters.reserve(theLowerDetSet.size());
+  if (theLowerDetSet.size() > 1)
+    LogDebug("VectorHitBuilderAlgorithm") << " more than 1 lower cluster! " << std::endl;
+  if (theUpperDetSet.size() > 1)
+    LogDebug("VectorHitBuilderAlgorithm") << " more than 1 upper cluster! " << std::endl;
+  for (const_iterator cil = theLowerDetSet.begin(); cil != theLowerDetSet.end(); ++cil) {
+    Phase2TrackerCluster1DRef clusterLower = edmNew::makeRefTo(clusters, cil);
+    lowerClusters.push_back(*clusterLower);
+  }
+  return true;
+}
+
+bool VectorHitBuilderAlgorithm::checkClustersCompatibility(Local3DPoint& poslower,
+                                                           Local3DPoint& posupper,
+                                                           LocalError& errlower,
+                                                           LocalError& errupper) const {
+  return true;
+}
+
+//----------------------------------------------------------------------------
+//ERICA::in the DT code the global position is used to compute the alpha angle and put a cut on that.
+std::vector<std::pair<VectorHit, bool>> VectorHitBuilderAlgorithm::buildVectorHits(
+    const StackGeomDet* stack,
+    edm::Handle<edmNew::DetSetVector<Phase2TrackerCluster1D>> clusters,
+    const detset& theLowerDetSet,
+    const detset& theUpperDetSet,
+    const std::vector<bool>& phase2OTClustersToSkip) const {
+  std::vector<std::pair<VectorHit, bool>> result;
+  if (checkClustersCompatibilityBeforeBuilding(clusters, theLowerDetSet, theUpperDetSet)) {
+    LogDebug("VectorHitBuilderAlgorithm") << "  compatible -> continue ... " << std::endl;
+  } else {
+    LogTrace("VectorHitBuilderAlgorithm") << "  not compatible, going to the next cluster";
+  }
+
+  std::vector<Phase2TrackerCluster1DRef> lowerClusters;
+  for (const_iterator cil = theLowerDetSet.begin(); cil != theLowerDetSet.end(); ++cil) {
+    Phase2TrackerCluster1DRef clusterLower = edmNew::makeRefTo(clusters, cil);
+    lowerClusters.push_back(clusterLower);
+  }
+  std::vector<Phase2TrackerCluster1DRef> upperClusters;
+  for (const_iterator ciu = theUpperDetSet.begin(); ciu != theUpperDetSet.end(); ++ciu) {
+    Phase2TrackerCluster1DRef clusterUpper = edmNew::makeRefTo(clusters, ciu);
+    upperClusters.push_back(clusterUpper);
+  }
+
+  std::sort_heap(
+      lowerClusters.begin(), lowerClusters.end(), LocalPositionSort(&*theTkGeom, &*theCpe, &*stack->lowerDet()));
+  std::sort_heap(
+      upperClusters.begin(), upperClusters.end(), LocalPositionSort(&*theTkGeom, &*theCpe, &*stack->upperDet()));
+
+  for (const auto& cluL : lowerClusters) {
+    LogDebug("VectorHitBuilderAlgorithm") << " lower clusters " << std::endl;
+    printCluster(stack->lowerDet(), &*cluL);
+    const PixelGeomDetUnit* gduLow = dynamic_cast<const PixelGeomDetUnit*>(stack->lowerDet());
+    auto&& lparamsLow = theCpe->localParameters(*cluL, *gduLow);
+    for (const auto& cluU : upperClusters) {
```

Comment:
Ok, at a later stage this full combinatorial loop need to be optmized
(first find first&last using some reasonable window)

Reply:
btw, being fully combinatorial: why sorting?

Reply:
actually from this code
https://cmssdt.cern.ch/lxr/source/RecoTracker/MeasurementDet/plugins/TkPhase2OTMeasurementDet.cc#0075
it seems that the clusters ARE sorted.
So either this assumption is wrong, or there is no need to sort.
btw the code above can be used also in here, once a window (pt-cut dependent) is defined.

Reply:
Got rid of the sorting.